### PR TITLE
feat(ai-promp-guard) add config option to prompt guard for scanning "system" and "assistant" messages

### PR DESCRIPTION
**Pull request by fffonion**
### Summary

This PR also fixes an issue when allow_all_conversation_history
is set to false, the first user request is selected instead of the last
one.

Unit tests are refactored to so that it's cleaner and has wider coverage.

### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

AG-58

### CODE REVIEW

**Review by jschmid1 — CHANGES_REQUESTED** [review]

File: kong/plugins/ai-prompt-guard/schema.lua [hunk 1]
```diff
@@ -32,13 +32,22 @@ return {
               type = "boolean",
               required = true,
               default = false } },
+          { match_all_roles = {
```
[the remark this answers is not in the document]
Comment (fffonion) [reply kong/plugins/ai-prompt-guard/schema.lua:35]: indeed, just fixed!

**Comment by fffonion** [comment]
@jschmid1 could you take another look on this? Thanks!

File: kong/plugins/ai-prompt-guard/schema.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (jschmid1) [reply kong/plugins/ai-prompt-guard/schema.lua:35]: This needs a test to verify backwards compat as-well :(

**Comment by jschmid1** [comment]
If you add a test for backwards compat verification, we can move this forward. In the meanwhile, I'll request another reviewer @Kong/gateway-moderators

File: kong/plugins/ai-prompt-guard/schema.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (fffonion) [reply kong/plugins/ai-prompt-guard/schema.lua:35]: TIL we have such tests, adding one now.

[the remark this answers is not in the document]
Comment (fffonion) [reply kong/plugins/ai-prompt-guard/schema.lua:35]: done

File: kong/plugins/ai-prompt-guard/handler.lua [hunk 2]
```diff
@@ -26,16 +26,21 @@ local execute do
   -- @tparam table request The deserialized JSON body of the request
   -- @tparam table conf The plugin configuration
   -- @treturn[1] table The decorated request (same table, content updated)
-  -- @treturn[2] nil
```
Comment (Tieske) [inline kong/plugins/ai-prompt-guard/handler.lua:29]: This line shouldn't be removed, it breaks the docs format

**Review by Tieske — CHANGES_REQUESTED** [review]

File: kong/plugins/ai-prompt-guard/handler.lua (hunk 2 above)
Comment (fffonion) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: there's another `treturn[2]` below, this is a typo i think

Comment (Tieske) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: each line is a (typed) value, the numbers indicate the return "options";

1. a valid result, 1 value, being a table
2. an error which returns 2 values:
    - a nil value
    - an error message

so 3 lines in total

see first example here: https://lunarmodules.github.io/ldoc/examples/multiple.lua.html

Comment (fffonion) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: I see, in other places however we use following patterns:
```lua
-- @treturn table          something
-- @treturn string|nil    error
```
should we align with that instead?

**Review by jschmid1 — APPROVED** [review]
👍🏼

File: kong/plugins/ai-prompt-guard/handler.lua (hunk 2 above)
Comment (fffonion) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: updated with `@treturn string|nil`

Comment (fffonion) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: There's PDK and other plugins uses this pattern, a took a snippet below:
```
kong/db/schema/topological_sort.lua:-- @treturn nil|string nil if the schemas were sorted, or a message if a cycle was found
kong/db/declarative/init.lua:-- @treturn nil|string error message, only if error happened
kong/db/declarative/init.lua:-- @treturn nil|table err_t, only if error happened
kong/db/declarative/init.lua:-- @treturn nil|string error message, only if error happened
kong/db/declarative/init.lua:-- @treturn nil|table err_t, only if error happened
kong/db/declarative/init.lua:-- @treturn nil|string error message if error
kong/db/declarative/init.lua:-- @treturn nil|table err_t if error
kong/tools/stream_api.lua:-- @treturn nil|string error
kong/tools/stream_api.lua:-- @treturn nil|string error   an error string
kong/tools/stream_api.lua:-- @treturn nil|string error  an error string, returned for protocol or socket I/O failures
kong/tools/stream_api.lua:-- @treturn nil|string   error
kong/tools/mime_type.lua:-- @treturn nil|nil|nil Invalid mime-type
kong/workspaces/counters.lua:-- @treturn nil|string error
kong/openid-connect/claims.lua:-- @return status(nil|boolean), error(string)
kong/openid-connect/claims.lua:-- @return status(nil|boolean), error(string)
kong/openid-connect/claims.lua:-- @return status(nil|boolean), error(string)
kong/openid-connect/claims.lua:-- @return status(nil|boolean), error(string)
kong/pdk/client/tls.lua:  -- @treturn nil|err Returns `nil` if successful, or an error message if it fails.
kong/pdk/client/tls.lua:  -- @treturn nil|err Returns `nil` if successful, or an error message if it fails.
kong/pdk/client/tls.lua:  -- @treturn nil|err Returns `nil` if successful, or an error message if it fails.
kong/pdk/request.lua:  -- @treturn nil|string An error message.
kong/pdk/client.lua:  -- @treturn nil|err `nil` if successful, or an error message if it fails.
kong/pdk/client.lua:  -- @treturn nil|err `nil` if successful, or an error message if it fails.
```

Right now, only the newly added ai-prompt-guard and ai-azure-content-safety plugin uses the new pattern. I'm not sure how we have two different patterns and our docs are rendering fine with the PDK ldoc.
Anyway, I'm going to just revert this change. I do not use ldoc locally so I don't have clue how to verify it.

Comment (Tieske) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: I don't know who wrote that, but most of that is wrong.

Comment (fffonion) [reply kong/plugins/ai-prompt-guard/handler.lua:29]: would you take another look at this PR? @Tieske

**Review by oowl — APPROVED** [review]

**Review by Tieske — APPROVED** [review]

**Pull request merged**